Support for multiple file formats

syberant opened the pull request:
Changed the --sleep functionality to a --max-fps implementation

Added (basic) support for the following formats:
- Life 1.05
- Run Length Encoded
- Plaintext files

and included examples

splintah commented on src/lib/mod.rs:
```diff
@@ -1,15 +1,15 @@
     pub board: Vec<Vec<bool>>,
     pub width: usize,
     pub height: usize,
+    pub generation: u64,
```
I think keeping track of the generation is not necessary.

syberant replied:
It was mostly for debugging at first but I thought it might be worth keeping around as an option, let me know what you think about another flag `--show-generation` or another mechanism for giving a user this option.

splintah commented on src/lib/mod.rs:
```diff
@@ -27,7 +27,8 @@ impl GameOfLife {
             width: width as usize,
-            height: height as usize
+            height: height as usize,
+            generation: 0u64,
```
**If** we're keeping generation, this literal should just be `0` instead of `0u64`.

syberant replied:
Can't argue with that.

splintah commented on src/lib/mod.rs:
```diff
@@ -51,10 +52,11 @@ impl GameOfLife {
                 self.board[y][x] = false;
             }
         }
+        self.generation = 0u64;
```
The same for this literal (`0` instead of `0u64`).

syberant replied:
Same, can't argue with that.

splintah commented on src/window.rs:
```diff
@@ -46,7 +47,7 @@ fn main() {
     let mut cell_width: u32 = 10;
     let mut chance: u8 = 220;
-    let mut sleep: Option<u64> = None;
+    let mut max_fps: u64 = 10;
```
I'd rather call this `fps`, and also make the flag `--fps`.

syberant replied:
I agree, that's probably a better name.
BTW, I don't really understand the piston event loop, I know it also processes `Input` events but if I exclude them then still for every frame that gets drawn with `--max-fps 1` 5 generations pass. I guess `Loop` events aren't always for drawing, I don't understand all of it but I do know that the `draw2d` function really gets run `max_fps` times a second.

splintah commented on src/window.rs:
```diff
@@ -93,11 +94,23 @@ fn main() {
         .build()
         .unwrap();
+    
+    // Create font for displaying text.
```
I have two problems with this font creating:
1. It happens at runtime, and if you have moved the binary elsewhere, the font isn't in this folder (I think).
2. It's unnecessary to display the generation.

syberant replied:
I agree with your first problem but disagree with the second, I think it **can** be handy to see how many generations have passed. It could also be very unnecessary of course but I would like to see it as an option.

splintah commented on src/window.rs:
```diff
@@ -136,11 +149,18 @@ fn main() {
-            let sleep = std::time::Duration::from_millis(sleep);
-            std::thread::sleep(sleep);
-        }
+            text(
```
Again, I don't want to display the generation.

splintah commented:
Also, before I'm going to merge it, please run:
- `rustfmt`, so the code style is consistent;
- and `clippy`, and change errors if they improve readability or simplicity.

splintah commented:
Regarding the parsers, I think the structure isn't optimal. I'd suggest the following:
- `parsers/mod.rs`: `struct Parser`, that has a method something like `fn get_board<S: AsRef<str>>(filename: S) -> Vec<Vec<bool>>`. It must do all things now done in `GameOfLife::init_with_file` that involve parsing.
- `parsers/life_106.rs` and other parsers: export only one function like `parse_life_106<S: ToString>(s: S) -> Result<Vec<Vec<bool>>, E>` (where `E` is a `String` or what is convenient).
The functions that check what file type the given file is should move to `parsers/mod.rs`, possibly as a method on `Parser`.

splintah commented:
I also want the `examples` folder to be without subfolders, so every example just in the examples folder.

syberant commented:
Overview of the changes:
- changed `--max-fps` and `max_fps` to better names according to @splintah's advice
- moved examples out of subfolders, removed the duplicated ones
- changed `0u64` to `0` for `generations`
- removed code and font for displaying `generations` but left the counting system in place
- style improvements by `rustfmt`

I am working on further reorganising the parsing of files, let me know whether you like the [direction of development](https://github.com/splintah/game-of-life/pull/1/commits/a7635e9e7b5516b759fb50df3f029274ab4662dd).

BTW: My build of `clippy` failed.

splintah commented:
I might reconsider generation counting, but in a different PR then.

splintah commented:
I've now created a foundation of the design I'd like to have for the parsers in 5da55d0.
Implementation shouldn't take that much time.

syberant commented:
I see you only want a parser for the cell data, I was trying to implement a parser that allowed for additional information such as the name of a pattern, this [`Pattern`](https://github.com/splintah/game-of-life/pull/1/commits/a7635e9e7b5516b759fb50df3f029274ab4662dd) struct could in the future also be used to export files.

Your suggested design will do it's job perfectly but won't be as expandable and I quite liked the idea of having such options open for development. If you don't like it I will implement it according to your suggestion but I would ask you to reconsider this, in my opinion, more open approach.

splintah commented:
I don’t see any uses for this metadata, but if you want to add it, you can add a tracking issue to discuss design and an implementation PR.

syberant commented:
Ok, so:
- the file parsing is now according to your design
- the `generation` counting system is commented out
- I have run `rustfmt` and `clippy`

I will probably try to implement some of the ideas that I have proposed in separate pull requests.

splintah commented:
I've now cleaned up the code, and removed the generation comments since it doesn't take that much time to reimplement it.

splintah commented:
One thing I would like to be added in this PR is testing for the parser. I'll add the basic layout for the testing.
Would you like to add the tests or shall I do it?

syberant commented:
You can do it, I need to learn some more rust to understand what exactly is going on there.

splintah commented:
> My guess might not be far off though, since your Travis CI build failed.
That's because I added the test layout, but all tests contained `unimplemented!()`, so they all panicked.
